File: shell/mobile/native/GophicsMonitor.swift
// Reference iOS MonitorHost for gophics's shell/mobile live-capture bridge.
//
// Wire once at startup:  bridge.setMonitorHost(GophicsMonitor(bridge: bridge))
//
// This is the streaming counterpart to GophicsMedia: MediaHost records one clip
// and hands it back when it stops, while MonitorHost keeps an open stream the Go
// side analyses live. A tuner or a singing app needs the latter — the audio has
// to be current, not eventual.

import AVFoundation
import Mobile
import Foundation

final class GophicsMonitor: NSObject {
    private let bridge: MobileBridge
    private let engine = AVAudioEngine()
    private var active = false
    private var reqID = 0

    init(bridge: MobileBridge) {
        self.bridge = bridge
    }

    private func ui(_ f: @escaping () -> Void) { DispatchQueue.main.async(execute: f) }
}

extension GophicsMonitor: MobileMonitorHostProtocol {

    // MARK: Permission

    func authorizeMic(_ reqID: Int) {
        AVAudioSession.sharedInstance().requestRecordPermission { granted in
            self.ui { self.bridge.deliverPermission(reqID, granted: granted) }
        }
    }

    // MARK: Monitoring

    func startMonitoring(_ reqID: Int) {
        if active { stopMonitoring(reqID) }
        self.reqID = reqID

        let session = AVAudioSession.sharedInstance()
        do {
            // .measurement is the point of this whole configuration: it turns
            // off the input processing — AGC, noise suppression, the voice EQ —
            // that iOS otherwise applies. That processing is tuned to make
            // speech intelligible and it distorts the harmonic structure pitch
            // detection reads, so a tuner using the default mode reads a voice
            // it has already altered.
            try session.setCategory(.playAndRecord, mode: .measurement,
                                    options: [.defaultToSpeaker, .allowBluetooth])
            try session.setActive(true)
        } catch {
            ui { self.bridge.failMonitoring(reqID, msg: "audio session: \(error.localizedDescription)") }
            return
        }

        let input = engine.inputNode
        let format = input.inputFormat(forBus: 0)
        guard format.sampleRate > 0, format.channelCount > 0 else {
            ui { self.bridge.failMonitoring(reqID, msg: "no audio input available") }
            return
        }

        input.installTap(onBus: 0, bufferSize: 1024, format: format) { [weak self] buffer, _ in
            guard let self, self.active else { return }
            guard let channels = buffer.floatChannelData else { return }

            let frames = Int(buffer.frameLength)
            if frames == 0 { return }
            let chCount = Int(buffer.format.channelCount)

            // Downmix to mono: the analyser is mono, and taking only channel 0
            // would silently halve the level on a stereo input.
            var mono = [Float](repeating: 0, count: frames)
            for c in 0..<chCount {
                let src = channels[c]
                for i in 0..<frames { mono[i] += src[i] }
            }
            if chCount > 1 {
                let inv = 1 / Float(chCount)
                for i in 0..<frames { mono[i] *= inv }
            }

            // Little-endian float32 bytes: gomobile binds only byte slices, so
            // a [Float] parameter is not expressible across the boundary.
            //
            // Delivered straight from the render thread on purpose. This
            // touches only a mutex-guarded ring buffer on the Go side and runs
            // hundreds of times a second; hopping to the main queue would add
            // latency to the one path that must stay current.
            mono.withUnsafeBufferPointer { p in
                let data = Data(buffer: p)
                self.bridge.deliverMonitorFloat32(self.reqID, data: data)
            }
        }

        engine.prepare()
        do {
            try engine.start()
        } catch {
            input.removeTap(onBus: 0)
            ui { self.bridge.failMonitoring(reqID, msg: "engine: \(error.localizedDescription)") }
            return
        }

        active = true
        let rate = Int(format.sampleRate)
        ui { self.bridge.deliverMonitorReady(reqID, sampleRate: rate) }
    }

    func stopMonitoring(_ reqID: Int) {
        guard active else { return }
        active = false
        engine.inputNode.removeTap(onBus: 0)
        engine.stop()
        // Give the session back so other apps — and this app's own playback —
        // are not left in a record-configured route.
        try? AVAudioSession.sharedInstance().setActive(false, options: .notifyOthersOnDeactivation)
    }
}

File: examples/mirror/android/app/src/main/java/dev/gophics/mirror/GophicsMonitor.kt
// Reference Android MonitorHost for gophics's shell/mobile live-capture Bind.
//
// Wire once at startup:  Bind.setMonitorHost(GophicsMonitor(activity, bridge))
// and route the runtime-permission result to onPermissionResult().
//
// This is the streaming counterpart to GophicsMedia: MediaHost records one clip
// and hands it back when it stops, while MonitorHost keeps an open stream that
// the Go side analyses live. A tuner or a singing app needs the latter — the
// audio has to be current, not eventual.

package dev.gophics.mirror

import android.Manifest
import android.app.Activity
import android.content.pm.PackageManager
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.MediaRecorder
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.Process
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import bind.Bind
import bind.MonitorHost

class GophicsMonitor(
    private val activity: Activity,
) : MonitorHost {

    private val ui = Handler(Looper.getMainLooper())
    private fun ui(f: () -> Unit) = ui.post(f)

    @Volatile private var running = false
    private var worker: Thread? = null
    private var pendingPermReq = 0L

    // --- Permission ----------------------------------------------------------

    override fun authorizeMic(reqID: Long) {
        if (ContextCompat.checkSelfPermission(activity, Manifest.permission.RECORD_AUDIO)
            == PackageManager.PERMISSION_GRANTED
        ) {
            Bind.deliverPermission(reqID, true)
            return
        }
        pendingPermReq = reqID
        ActivityCompat.requestPermissions(
            activity, arrayOf(Manifest.permission.RECORD_AUDIO), REQ_RECORD_AUDIO
        )
    }

    /** Call from Activity.onRequestPermissionsResult for REQ_RECORD_AUDIO. */
    fun onPermissionResult(granted: Boolean) {
        if (pendingPermReq != 0L) {
            Bind.deliverPermission(pendingPermReq, granted)
            pendingPermReq = 0L
        }
    }

    // --- Monitoring ----------------------------------------------------------

    override fun startMonitoring(reqID: Long) {
        if (running) stopMonitoring(reqID)

        if (ContextCompat.checkSelfPermission(activity, Manifest.permission.RECORD_AUDIO)
            != PackageManager.PERMISSION_GRANTED
        ) {
            ui { Bind.failMonitoring(reqID, "microphone permission not granted") }
            return
        }

        val rec = openRecorder()
        if (rec == null) {
            ui { Bind.failMonitoring(reqID, "could not open the microphone") }
            return
        }

        running = true
        try {
            rec.startRecording()
        } catch (e: IllegalStateException) {
            running = false
            rec.release()
            ui { Bind.failMonitoring(reqID, "microphone busy: ${e.message}") }
            return
        }

        val rate = rec.sampleRate
        ui { Bind.deliverMonitorReady(reqID, rate.toLong()) }

        worker = Thread({
            // Audio priority: a scheduling stall here shows up as a dropped
            // buffer, which reads to the user as the pitch display stuttering.
            Process.setThreadPriority(Process.THREAD_PRIORITY_URGENT_AUDIO)
            val buf = ByteArray(READ_BYTES)
            while (running) {
                val n = rec.read(buf, 0, buf.size)
                if (n <= 0) {
                    // ERROR_INVALID_OPERATION / ERROR_DEAD_OBJECT: the device
                    // went away (a call came in, or the app was backgrounded).
                    if (n < 0) break
                    continue
                }
                // Deliberately NOT marshaled to the UI thread. This touches only
                // a mutex-guarded ring buffer on the Go side, runs hundreds of
                // times a second, and hopping threads would add latency to the
                // one path that must stay current. See MonitorHost's doc.
                Bind.deliverMonitorPCM(reqID, if (n == buf.size) buf else buf.copyOf(n))
            }
            try {
                rec.stop()
            } catch (_: IllegalStateException) {
                // Already stopped; releasing is what matters.
            }
            rec.release()
        }, "gophics-monitor").also { it.start() }
    }

    override fun stopMonitoring(reqID: Long) {
        running = false
        worker?.join(500)
        worker = null
    }

    /**
     * Opens an AudioRecord, preferring the least-processed input available.
     *
     * The source matters more than it looks. MIC runs the platform's voice
     * processing — automatic gain control, noise suppression, sometimes an
     * aggressive high-pass — all of which are tuned to make speech intelligible
     * over a phone line and all of which distort the harmonic structure that
     * pitch detection reads. UNPROCESSED asks for the raw capture path;
     * VOICE_RECOGNITION is the next best, since it also disables AGC on most
     * devices. MIC is the last resort.
     */
    private fun openRecorder(): AudioRecord? {
        val sources = mutableListOf<Int>()
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            sources.add(MediaRecorder.AudioSource.UNPROCESSED)
        }
        sources.add(MediaRecorder.AudioSource.VOICE_RECOGNITION)
        sources.add(MediaRecorder.AudioSource.MIC)

        for (source in sources) {
            for (rate in RATES) {
                val min = AudioRecord.getMinBufferSize(
                    rate, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT
                )
                if (min <= 0) continue
                val rec = try {
                    AudioRecord(
                        source, rate,
                        AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT,
                        maxOf(min, READ_BYTES * 4)
                    )
                } catch (_: IllegalArgumentException) {
                    continue
                }
                if (rec.state == AudioRecord.STATE_INITIALIZED) return rec
                rec.release()
            }
        }
        return null
    }

    companion object {
        const val REQ_RECORD_AUDIO = 0x6702

        /** Preferred capture rates, best first. */
        private val RATES = intArrayOf(48000, 44100, 32000, 16000)

        /**
         * Bytes per read — 1024 frames of 16-bit mono, about 21 ms at 48 kHz.
         * Small enough that the Go ring buffer stays current, large enough that
         * the JNI crossing is not the dominant cost.
         */
        private const val READ_BYTES = 2048
    }
}
